### ISSUE TITLE
[Bug] Alloy-sol-type-parser blocking compile of alloy crate

### ISSUE THREAD

**Issue by CRossel87a** [alloy-rs/core#717]
### Component

sol-type-parser

### What version of Alloy are you on?

0.3.0

### Operating System

Windows

### Describe the bug

   Compiling alloy-sol-type-parser v0.8.0
error[E0277]: the trait bound `CustomError: StdError` is not satisfied
  --> C:\Users\chris\.cargo\registry\src\index.crates.io-6f17d22bba15001f\alloy-sol-type-parser-0.8.0\src\input.rs:39:13 
   |
39 |             winnow::error::ErrMode::from_external_error(input, winnow::error::ErrorKind::Eof, err)
   |             ^^^^^^^^^^^^^^^^^^^^^^ the trait `StdError` is not implemented for `CustomError`, which is required by `ErrMode<_>: FromExternalError<_, _>`

Running latest Rust version

**Comment by CRossel87a** [comment]
Ah, this only happens if you have ethers in the same dependency list

**Comment by shuhuiluo** [comment]
This issue is also bugging me. I see "the trait `core::error::Error` is not implemented for `CustomError`, which is required by `ErrMode<_>: FromExternalError<_, _>`".

Can confirm removing `ethers` solves the problem. Or add
```rust
#[cfg(not(feature = "std"))]
impl core::error::Error for CustomError {}
```
to `alloy-sol-type-parser` `input.rs`.

**Comment by DaniPopes** [comment]
are you using no_std? you need to enable alloy-sol-type-parser/std feature if winnow/std is also enabled

**Comment by shuhuiluo** [comment]
> are you using no_std? you need to enable alloy-sol-type-parser/std feature if winnow/std is also enabled

I'm. But it doesn't compile with or without `std` feature.

**Comment by DaniPopes** [comment]
I dont believe that, the problem is a cfg(std) impl in winnow, you have to enable std in alloy as well for it to work

**Comment by shuhuiluo** [comment]
> I dont believe that, the problem is a cfg(std) impl in winnow, you have to enable std in alloy as well for it to work

@DaniPopes Can you take a look at [this branch](https://github.com/shuhuiluo/uniswap-v3-sdk-rs/tree/sol-type-parser-error)? I can only make it compile by removing `ethers` or `impl core::error::Error for CustomError`.

**Comment by Wollac** [comment]
I can confirm that I see this problem as well. It seems to be an issue with feature unification/deduplication. The build fails when `alloy` is used with certain other dependencies that use `toml_edit` or `winnow` respectively. It does not matter if the `std` flag is given or not.
The following is a minimal example that fails and only compiles if the other dependency is removed:
```toml
[dependencies]
alloy = { version = "0.3.0" , features = ["full"] }
borsh-derive = "1.5.1"
``` 
or as mentioned by @shuhuiluo `ethers`:
```toml
[dependencies]
alloy = { version = "0.3.0" , features = ["full"] }
ethers = "2.0.14"
```

**Comment by CRossel87a** [comment]
> > I dont believe that, the problem is a cfg(std) impl in winnow, you have to enable std in alloy as well for it to work
> 
> @DaniPopes Can you take a look at [this branch](https://github.com/shuhuiluo/uniswap-v3-sdk-rs/tree/sol-type-parser-error)? I can only make it compile by removing `ethers` or `impl core::error::Error for CustomError`.

Does not compile with stable:
error[E0658]: use of unstable library feature 'error_in_core'
  --> \crates\sol-type-parser\src\input.rs:28:6
   |
28 | impl core::error::Error for CustomError {}

**Comment by CRossel87a** [comment]
I found a short term workaround for devs. This set of dependencies will compile:

[dependencies]
alloy = { version = "0.2.1", features = ["full"] }
ethers = "2.0.14"
winapi = { version = "0.3.9", features = ["winerror"] }

**Comment by Wollac** [comment]
Downgrading `alloy` works, but this issue should also be fixed for the latest version.

**Comment by DaniPopes** [comment]
I have no idea what is going on here, because just importing alloy-sol-type-parser alone, or with winnow, with any feature combination works fine, but the moment you import alloy with full and num_enum it fails

**Comment by Wollac** [comment]
It's a very strange issue indeed, and I thought I was going crazy trying to track it down.
The only two options I can think of right now are to a) remove the `no_std` support entirely or b) find an alternative to #718 that does not break the MSRV.

**Comment by CRossel87a** [comment]
cheers

**Comment by robriks** [comment]
Another workaround is to avoid the use of Alloy-sol-type-parser by using the sol macro's syn-solidity parser instead of the solc standard json abi approach

**Comment by DaniPopes** [comment]
Those are two separate parsers for two different purposes, don't know what you're trying to achieve; alloy-sol-type-parser is used internally and the issue is fixed in >=0.8.2
